### ISSUE TITLE
Add BF16-native and mixed-precision YatNMN execution modes

### ISSUE THREAD

**Issue by mlnomadpy** [azettaai/nmn#47]
## Motivation

`YatNMN` currently promotes the complete input and kernel operands to FP32 in both the standard and fused Flax NNX paths:

```python
inputs_f32 = inputs.astype(jnp.float32)
kernel_f32 = kernel.astype(jnp.float32)
```

This is a safe reference implementation, but it prevents an actually BF16-native Yat FFN and may add conversion traffic or FP32 operand materialization on TPU. This becomes important when `YatNMN` replaces the FFN in a transformer, where the operands are much larger than in the existing small examples.

The accompanying comment says that squaring the score causes BF16 overflow. BF16 has the same exponent width and essentially the same dynamic range as FP32, so overflow is not normally the distinguishing failure mode. The important BF16 risks here are instead:

1. the seven-bit mantissa;
2. cancellation in
   `||x||² + ||w||² - 2 xᵀw` when `x` and `w` are close;
3. error amplification by the reciprocal denominator; and
4. the `D^{-2}` dependence appearing in derivatives of the rational score.

Full FP32 evaluation addresses all four, but it is more conservative than necessary for every workload.

## Proposed design

Preserve the current FP32 implementation as the default/reference path, and add explicit opt-in numerical modes, for example:

```python
YatNMN(
    ...,
    compute_mode="fp32",       # "fp32" | "mixed" | "bf16"
    distance_floor=0.0,
)
```

The exact API is open for discussion; the important point is that `dtype=jnp.bfloat16` currently does not imply BF16 Yat-score evaluation because the layer overrides it internally.

### 1. `fp32`: reference behavior

Keep the existing implementation and cross-framework FP32 parity unchanged.

### 2. `mixed`: recommended TPU path

Keep parameters and activations as BF16 and pass BF16 operands directly to `dot_general`, while requesting FP32 accumulation/output with `preferred_element_type=jnp.float32` when supported. Norm reductions can similarly use an FP32 accumulator without first materializing complete FP32 copies of the input and kernel:

```python
dot = lax.dot_general(
    x_bf16,
    w_bf16,
    dimension_numbers,
    precision=precision,
    preferred_element_type=jnp.float32,
)
x_sq = jnp.sum(x_bf16 * x_bf16, axis=-1, keepdims=True, dtype=jnp.float32)
w_sq = jnp.sum(w_bf16 * w_bf16, axis=0, keepdims=True, dtype=jnp.float32)
```

This should retain the numerically sensitive accumulations while avoiding the unconditional full-operand casts. Whether XLA already fuses the current conversions should be verified from lowered HLO and TPU profiles rather than assumed.

### 3. `bf16`: experimental strict-BF16 path

Use a dimension-scaled form so intermediates stay near unit scale. For input width `d`, define

```text
m_j  = mean_i(x_i w_ij)
r_x  = mean_i(x_i²)
r_wj = mean_i(w_ij²)
r_j  = max(r_x + r_wj - 2 m_j, δ_j)
```

Then compute

```text
y_j = d (m_j + b_j/d)² / (r_j + ε/d)
```

followed by the existing alpha scaling. Before the distance floor and finite-precision rounding, this is algebraically identical to

```text
(xᵀw_j + b_j)² / (||x - w_j||² + ε).
```

`δ_j` could be an absolute floor or a relative floor such as

```text
max(abs_floor, rel_floor * (r_x + r_wj)).
```

The spherical path needs the analogous protection around `2 - 2 xᵀw`. The fused/custom-VJP implementation should differentiate the clamped forward expression consistently rather than silently using an unrelated gradient rule.

Input RMS normalization, controlled kernel norms, and global gradient clipping remain useful training-level safeguards, but they should not be required to make the layer numerically defined.

## Validation plan

Please benchmark all three modes rather than replacing the reference implementation immediately:

- input widths from 128 to 4096;
- random inputs plus adversarial near-collision cases `x ≈ w`;
- several input/kernel scales and epsilon values from `1e-5` to `1`;
- forward absolute/relative error against FP32;
- input/kernel gradient error and gradient cosine similarity;
- NaN/Inf incidence and maximum activation/gradient magnitude;
- standard versus fused path parity;
- lowered HLO, peak memory, compile time, and steady-state step time on TPU;
- a 500–1000-step transformer FFN training smoke test.

Suggested acceptance criteria:

1. `compute_mode="fp32"` remains numerically equivalent to the current release.
2. `mixed` avoids explicit full-array FP32 operand casts and stays close to the FP32 forward and backward reference.
3. `bf16` remains finite across the declared scale/epsilon test domain, with its error envelope documented.
4. Precision behavior is documented for both standard and fused NNX paths.

This would let users choose between strict reference accuracy, a likely best TPU performance/accuracy compromise, and an experimental all-BF16 implementation without changing the current default silently.
